src/lambertw.jl

```diff
@@ -0,0 +1,287 @@
+    if k == 0
+        x == 0 && return float(zero(x))
+        x == 1 && return convert(typeof(float(x)), LambertW.Omega) # must be a more efficient way
+    end
+    return _lambertw(float(x), k, maxits)
+end
+
+### Real x
+
+_lambertw(x::Real, k::Integer, maxits::Integer) = _lambertw(x, Val(Int(k)), maxits)
+
+# Real x, k = 0
+# There is a magic number here. It could be noted, or possibly removed.
+# In particular, the fancy initial condition selection does not seem to help speed.
+function _lambertw(x::T, ::Val{0}, maxits::Integer) where T<:Real
+    isfinite(x) || return x
+    one_t = one(T)
+    oneoe = -T(inve)  # The branch point
+    x == oneoe && return -one_t
+    oneoe < x || throw(DomainError(x))
+    itwo_t = 1 / convert(T, 2)
+    if x > one_t
+        lx = log(x)
+        llx = log(lx)
+        x0 = lx - llx - log(one_t - llx / lx) * itwo_t
+    else
+        x0 = (567//1000) * x
+    end
+    return lambertw_root_finding(x, x0, maxits)
+end
+
+# Real x, k = -1
+function _lambertw(x::T, ::Val{-1}, maxits::Integer) where T<:Real
+    oneoe = -T(inve)
+    x == oneoe && return -one(T) # W approaches -1 as x -> -1/e from above
+    oneoe < x || throw(DomainError(x))  # branch domain exludes x < -1/e
+    x == zero(T) && return -convert(T, Inf) # W decreases w/o bound as x -> 0 from below
+    x < zero(T) || throw(DomainError(x))
+    return lambertw_root_finding(x, log(-x), maxits)
+end
+
+_lambertw(x::Real, k::Val, maxits::Integer) =
+    throw(DomainError(x, "lambertw: for branch k=$k not defined, real x must have branch k == 0 or k == -1"))
+
+### Complex z
+
+_lambertw(z::Complex{<:Integer}, k::Integer, maxits::Integer) = _lambertw(float(z), k, maxits)
+# choose initial value inside correct branch for root finding
+function _lambertw(z::Complex{T}, k::Integer, maxits::Integer) where T<:Real
+    local w::Complex{T}
+    pointseven = 7//10
+    if abs(z) <= T(inve)
+        if z == 0
+            k == 0 && return z
+            return complex(-convert(T, Inf), zero(T))
+        end
+        if k == 0
+            w = z
+        elseif k == -1 && imag(z) == 0 && real(z) < 0
+            w = complex(log(-real(z)), 1//10^7) # need offset for z ≈ -1/e.
+        else
+            w = log(z)
+            k != 0 ? w += complex(0, k * 2 * pi) : nothing
+        end
+    elseif k == 0 && imag(z) <= pointseven && abs(z) <= pointseven
+        w = abs(z+ 1//2) < 1//10 ? imag(z) > 0 ?
+                complex(pointseven, pointseven) :
+                complex(pointseven, -pointseven) : z
+    else
+        if real(z) == convert(T, Inf)
+            k == 0 && return z
+            return z + complex(0, 2*k*pi)
+        end
+        real(z) == -convert(T, Inf) && return -z + complex(0, (2*k+1)*pi)
+        w = log(z)
+        k != 0 ? w += complex(0, 2*k*pi) : nothing
+    end
+    return lambertw_root_finding(z, w, maxits)
+end
+
+### root finding, iterative solution
+
+# Use Halley's root-finding method to find
```

Comment:
Worth noting that there's a paper from Lóczi published last year in Applied Mathematics and Computation called "Guaranteed- and high-precision evaluation of the Lambert W function" that describes algorithms that improve significantly on e.g. Halley's method. I used it to implement _W_<sub>-1</sub> for real arguments in (-1/_e_, 0) in https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/lindley.jl#L137-L156.

Reply:
I've actually tried the iteration scheme evaluated in the paper, but it looks like it's on par with Halley or the latter requires 1-2 iterations less from the same starting point (but I have not tested very extensively).
IIUC the paper doesn't claim that the Iacono-Boyd method provides faster convergence than Halley method, it just proves the bounds for the Iacono-Boyd one (because it's a simpler formula).

Reply:
Ah good to know, it's been a bit since I read it so the details have been largely GC'd. Probably should have reread the paper before attempting to summarize it. 😅 One thing I recall was that when I was testing things out for the -1 branch in (-1/e, 0), I was comparing LambertW.jl vs. Iacono-Boyd vs. Mathematica as "ground truth" and found that Iacono-Boyd produced more accurate results near -1/e and 0 but was otherwise the same. Your comment makes me realize that I had actually neglected to check the number of iterations that LambertW.jl required for convergence. LambertW.jl explicitly keeps track of the change between iterations and will stop early based on that, whereas my naive implementation of Iacono-Boyd just iterates a set number of times. In developing it, I was playing around with the number of iterations and found that most inputs required only 2 or so iterations to converge, so I'd be interested to see how that compares with LambertW.jl.

Reply:
I've tried to implement Iacono-Boyd scheme here: https://github.com/alyst/SpecialFunctions.jl/tree/lambertw_iacono_iters.
Actually, the iteration scheme is just an input parameter to the root finding function, so that code could be easily adopted for more comprehensive evaluation of efficiency of both schemes over the whole *LambertW(x)* domain and for different numeric types.